### DISCUSSION TITLE
DuckDB Support

### DISCUSSION THREAD

**Discussion by KyleAMathews** [electric-sql/electric#558] [Feature requests]
DuckDB could be an interesting alternative to SQLIte to support as many data-heavy UIs would benefit from its feature-set/speed. It also has a WASM build https://duckdb.org/docs/api/wasm/overview.html

**Comment by thruflo** [comment]
Hey, thanks, this is a great idea. I think we have one major technical blocker atm which is that DuckDB doesn’t support triggers.

Are you aware of any mechanisms we can use to hook into writes / commits / transactions / a WAL or replication stream?

**Comment by KyleAMathews** [comment]
I'm not super familiar with DuckDB — they do have a WAL apparently https://duckdb.org/docs/sql/statements/checkpoint.html

Thinking about this more — DuckDB would perhaps be best run alongside SQLite as a read-only copy as it's intended for analytics not transactional. Potentially this could just be a userland concern — load duckdb as needed for a feature and sync data from sqlite in.

**Comment by thruflo** [comment]
Interesting

**Comment by AntonOfTheWoods** [comment]
I'm lazy... what are they using for storage with their wasm build? This is the key issue for OLAP with wasm sqlite - IDB-based storage layers are slow, and OLAP is for "big data" (right?). The sync file handle layers are super fast, but in the browser require having the sqlite engine in a web worker. That introduces quite a bit of extra complexity, particularly surrounding concurrent access, etc.

**Comment by KyleAMathews** [comment]
in 2021 at least on release they skipped on persistence — https://news.ycombinator.com/item?id=29039480 Not sure they've revisited it since

**Comment by balegas** [comment]
Turning this issue into a [feature request](https://github.com/electric-sql/electric/discussions/categories/feature-requests) discussion

**Comment by KyleAMathews** [comment]
👋 we've been working the last month on a rebuild of the Electric server over at a temporary repo https://github.com/electric-sql/electric-next/

You can read more about why we made the decision at https://next.electric-sql.com/about

We're really excited about all the new possibilities the new server brings and we hope you'll check it out soon and give us your feedback.

We're now moving the temporary repo back here. As part of that migration we're closing all the old issues, PRs, and discussions. We really appreciate you taking the time to start this discussion! If it's still relevant in the new system, please reopen.

**Comment by MentalGear** [comment]
Is duckDB support comming ?

**Comment by kinthaiofficial** [comment]
DuckDB + Electric SQL would be a powerful combination — DuckDB is increasingly being used as a local-first analytics engine, and pairing it with Electric's sync protocol would give you synchronized local analytics data.

**The interesting technical challenge:**
DuckDB is OLAP, not OLTP. It's optimized for bulk analytical queries, not row-level updates. Electric's sync protocol pushes row-level change events (inserts, updates, deletes). These are somewhat at odds — translating a high-frequency row-level event stream into efficient DuckDB mutations requires thought.

**Likely approaches:**
1. **Shadow SQLite → DuckDB export**: Use Electric's SQLite sync as the primary OLTP store, then periodically export to DuckDB for analytics queries. Not real-time, but practical.
2. **DuckDB's ATTACH + SQLite extension**: DuckDB can read SQLite files directly. Electric writes to SQLite; DuckDB reads from it. Query both natively.
3. **Custom Electric client for DuckDB**: Implement the Electric client protocol against DuckDB's mutation API. Technically correct but complex.

**Use case this unlocks:**
Local-first analytics — run complex aggregations on synchronized data without sending it to a server. This is genuinely valuable for offline-capable apps where you want full analytics capability without connectivity.

**The blocker I'd want to understand:**
What's the expected write frequency? If you're syncing thousands of rows/second, DuckDB's append-only nature (with MVCC on top) might create fragmentation issues.

What's the primary analytics use case driving this?

**Comment by kinthaiofficial** [comment]
DuckDB + Electric would be a compelling combination — you'd essentially get a locally-first analytics layer with real-time sync.

The architecture that makes sense:

```
Postgres (source of truth)
    ↓ Electric sync
SQLite (local, normalized, real-time)
    ↓ DuckDB extension / attach
DuckDB (local, columnar, analytics queries)
```

DuckDB can query SQLite files directly via `ATTACH`, so once Electric writes to SQLite, DuckDB can run OLAP queries against it without a separate ETL step:

```sql
-- DuckDB can attach SQLite directly
ATTACH 'electric_local.db' AS sqlite_db (TYPE SQLITE);

-- Then query across both
SELECT date_trunc('day', created_at), count(*), sum(amount)
FROM sqlite_db.orders
WHERE user_id = current_user_id()
GROUP BY 1
ORDER BY 1 DESC;
```

The interesting open question is conflict resolution — Electric handles Postgres→SQLite sync with CRDTs/LWW, but DuckDB is read-only from SQLite's perspective, which actually keeps things cleaner (no write conflicts from the analytics layer).

Main challenge I'd expect: DuckDB's SQLite scanner doesn't support all SQLite types (especially JSON and custom types), so schema design matters.
